src/Xamarin.Android.Build.Tasks/Utilities/Aapt2Daemon.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using Microsoft.Build.Framework;
+
+namespace Xamarin.Android.Tasks {
+	class Aapt2Daemon : IDisposable {
+		struct DaemonProcess {
+			public Process Process;
+			public object LockObject;
+		}
+		ConcurrentDictionary<int, DaemonProcess> daemons = new ConcurrentDictionary<int, DaemonProcess> ();
+
+		public string Aapt2 { get; private set; }
+
+		public string ToolName { get { return Path.GetFileName (Aapt2); }}
+
+		public static Aapt2Daemon GetInstance (IBuildEngine4 engine, string aapt2) {
+			Aapt2Daemon daemon = (Aapt2Daemon)engine.GetRegisteredTaskObject (typeof(Aapt2Daemon).FullName,  RegisteredTaskObjectLifetime.Build);
+			if (daemon == null)
+			{
+				daemon = new Aapt2Daemon (aapt2);
+				engine.RegisterTaskObject (typeof (Aapt2Daemon).FullName, daemon, RegisteredTaskObjectLifetime.Build, allowEarlyCollection: false);
+			}
+			return daemon;
+		}
+
+		public Aapt2Daemon (string aapt2)
+		{
+			Aapt2 = aapt2;
+		}
+
+		/*
+		`aapt2 daemon` allows commands to be passed via the StdIn Stream.
+		However each "argument" or "switch" needs to be on its own line.
+		so 
+			'--manifest AndroidManifest.xml'
+		will NOT work
+			`--manifest`
+			`AndroidManifest.xml`
+		does work.
+
+		This needs to be done for ALL arguments. 
+		*/
+		bool Run (DaemonProcess aapt2, IList<OutputLine> output, params string[] args) {
+			lock (aapt2.LockObject) {
+				StreamWriter writer = aapt2.Process.StandardInput;
+				Console.WriteLine ($"Writer = {writer.Encoding}");
+				foreach (var arg in args) {
+					writer.WriteLine (arg.Trim ());
+				}
+				writer.WriteLine ();
+				writer.Flush ();
+			}
+			string line;
+			bool errored = false;
+			while ((line = aapt2.Process.StandardError.ReadLine ()) != null) {
+				if (line == "Done")
+					break;
+				if (line == "Error") {
+					errored = true;
+					break;
+				}
+				output.Add (new OutputLine (line, stdError: !IsAapt2Warning (line)));
+			}
+			return !errored;
+		}
+
+		public bool RunAapt2 (IList<OutputLine> output, params string [] args)
+		{
+			int id = Thread.CurrentThread.ManagedThreadId;
+			if (daemons.TryGetValue (id, out DaemonProcess aapt2)) {
+				return Run(aapt2, output, args);
+			}
+			ProcessStartInfo info = new ProcessStartInfo (Aapt2) {
+				Arguments = "daemon",
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+				RedirectStandardInput = true,
+				RedirectStandardError = true, 
+				RedirectStandardOutput = true,
+				UseShellExecute = false,
+				StandardErrorEncoding = Encoding.UTF8,
+				StandardOutputEncoding = Encoding.UTF8,
+				//StandardInputEncoding = Encoding.UTF8,
+			};
+			aapt2.LockObject = new object ();
+			aapt2.Process = new Process ();
+			aapt2.Process.StartInfo = info;
+			aapt2.Process.Start();
+			if (daemons.TryAdd (id, aapt2)) {
```

Comment:
Should this use [GetOrAdd](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.8)?

If this finds a key that is already present, it wouldn't even call aapt2. Is that correct or not?